File: 1095-Find_in_Mountain_Array/Find_in_Mountain_Array.java
/**
 * // This is MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface MountainArray {
 *     public int get(int index) {}
 *     public int length() {}
 * }
 */
 
class Solution {
    HashMap<Integer, Integer> cache = new HashMap<>();
    MountainArray mountainArr;
    
    public int findInMountainArray(int target, MountainArray mountainArr) {
        this.mountainArr = mountainArr;
        int n = mountainArr.length();
        
        int left = 1;
        int right = n - 2;
        
        // To find the peak
        while(left != right){
            int mid = (left + right)/2;
            if(getValue(mid) < getValue(mid+ 1)){
                left = mid + 1;
            }
            else {
                right = mid;
            }
        }
        
        int peak = right;
        
        // Checking the left side of the mountain array
        left = 0;
        right = peak;
        while(left <= right){
            int mid = left + (right - left)/2;
            int midValue = getValue(mid);
            if(midValue == target){
                return mid;
            }
            else if(midValue < target){
                left = mid + 1;
            }
            else{
                right = mid - 1;
            }
        }
        
        // Checking the right side of the mountain array
        left = peak + 1;
        right = n - 1;
        while(left <= right){
            int mid = left + (right - left)/2;
            int midValue = getValue(mid);
            if(midValue == target){
                return mid;
            }
            else if(midValue > target){
                left = mid + 1;
            }
            else{
                right = mid - 1;
            }
        }
        
        return -1;
    }
    
    private int getValue(int idx){
        if(cache.containsKey(idx)){
            return cache.get(idx);
        }
        
        int value = mountainArr.get(idx);
        cache.put(idx, value);
        return value;
    }
}

File: 1095-FindInMountainArray/FindInMountainArray.go
package findinmountainarray

// Dummy MountainArray implementation for compilation
// Actual implementation of Mountain Array shoudl be done on LC website
type MountainArray struct {
	arr []int
}

func (this *MountainArray) get(index int) int {
	return this.arr[index]
}

func (this *MountainArray) length() int {
	return len(this.arr)
}

var cache map[int]int

func findInMountainArray(target int, mountainArr *MountainArray) int {
	cache = make(map[int]int)
	n := mountainArr.length()

	left := 1
	right := n - 2

	// Finding the peak idx
	for left != right {
		mid := (left + right) >> 1

		if getValue(mid, mountainArr) < getValue(mid+1, mountainArr) {
			left = mid + 1
		} else {
			right = mid
		}
	}

	peak := right

	// Searching the left side of the mountain array
	left = 0
	right = peak

	for left <= right {
		mid := (left + right) >> 1
		midValue := getValue(mid, mountainArr)

		if midValue == target {
			return mid
		} else if midValue < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	// Searching the right side of the mountain array
	left = peak
	right = n - 1

	for left <= right {
		mid := (left + right) >> 1
		midValue := getValue(mid, mountainArr)

		if midValue == target {
			return mid
		} else if midValue > target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	// Target not found
	return -1
}

func getValue(idx int, mountainArr *MountainArray) int {
	val, exists := cache[idx]
	if exists {
		return val
	}

	// Retrieve and insert into the cache
	val = mountainArr.get(idx)
	cache[idx] = val
	return val
}
